spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/core/mapping/SpannerObjectMapperImpl.java

```diff
@@ -0,0 +1,336 @@
+/*
+ *  Copyright 2018 original author or authors.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package org.springframework.cloud.gcp.data.spanner.core.mapping;
+
+import java.lang.reflect.Array;
+import java.lang.reflect.Constructor;
+import java.lang.reflect.Method;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.BiConsumer;
+import java.util.function.BiFunction;
+import java.util.function.Consumer;
+import java.util.function.Predicate;
+
+import com.google.cloud.ByteArray;
+import com.google.cloud.Date;
+import com.google.cloud.Timestamp;
+import com.google.cloud.spanner.Mutation.WriteBuilder;
+import com.google.cloud.spanner.ResultSet;
+import com.google.cloud.spanner.Struct;
+import com.google.cloud.spanner.Type;
+import com.google.cloud.spanner.ValueBinder;
+import com.google.common.collect.ImmutableMap;
+
+import org.springframework.data.mapping.PersistentPropertyAccessor;
+import org.springframework.data.mapping.PropertyHandler;
+import org.springframework.util.Assert;
+
+/**
+ * @author Chengyuan Zhao
+ */
+public class SpannerObjectMapperImpl implements SpannerObjectMapper {
+
+	private final SpannerMappingContext spannerMappingContext;
+
+	// used for caching the methods used for reading to entities. Does not contain the
+	// methods for
+	// properties with inner types, such as Iterable.
+	private final Map<Class, Method> propertyReadMethodMapping = new HashMap<>();
+
+	// used for caching the methods used for writing to mutations. Does not contain the
+	// methods for
+	// properties with inner types, such as Iterable.
+	private final Map<Class, Method> propertyWriteMethodMapping = new HashMap<>();
+
+	private static final Map<Class, BiConsumer<ValueBinder<WriteBuilder>, Iterable>>
+			writeBuilderIterableMapping =
+			// The casting of each biconsumer below is needed for Java 8 compilation, but not Java 9.
+			new ImmutableMap.Builder<Class, BiConsumer<ValueBinder<WriteBuilder>, Iterable>>()
+			.put(Boolean.class,
+					(BiConsumer<ValueBinder<WriteBuilder>, Iterable>) (binder,
```

Comment:
This compilation error is so strange...
works for me when I add the elements one by one using the builder reference:  

```
	private static final Map<Class, BiConsumer<ValueBinder<WriteBuilder>, Iterable>>
			writeBuilderIterableMapping =
			// The casting of each biconsumer below is needed for Java 8 compilation, but not Java 9.
			createWriteBuilderIterableMapping();

	private static ImmutableMap<Class, BiConsumer<ValueBinder<WriteBuilder>, Iterable>> createWriteBuilderIterableMapping() {
		Builder<Class, BiConsumer<ValueBinder<WriteBuilder>, Iterable>> builder = new Builder<>();

		builder.put(Date.class, ValueBinder::toDateArray);
		builder.put(Boolean.class, ValueBinder::toBoolArray);
		builder.put(Long.class, ValueBinder::toInt64Array);
		builder.put(String.class, ValueBinder::toStringArray);
		builder.put(Double.class, ValueBinder::toFloat64Array);
		builder.put(Timestamp.class,ValueBinder::toTimestampArray);
		builder.put(ByteArray.class, ValueBinder::toBytesArray);

		return builder.build();
	}
```